test/e2e/cleanup/iam.go

```diff
@@ -0,0 +1,176 @@
+package cleanup
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	"github.com/aws/aws-sdk-go-v2/service/iam/types"
+	"github.com/go-logr/logr"
+
+	"github.com/aws/eks-hybrid/test/e2e/constants"
+	"github.com/aws/eks-hybrid/test/e2e/errors"
+)
+
+type IAMCleaner struct {
+	IAM    *iam.Client
+	Logger logr.Logger
+}
+
+func (c *IAMCleaner) ListRoles(ctx context.Context, filterInput FilterInput) ([]string, error) {
+	var roles []string
+
+	// list-roles does not allow filtering by tags so we have to pull them all
+	// We have the role =* checks to try and limit which roles we bother checking tags for
+	// but we only delete those with the e2e cluster tag
+	paginator := iam.NewListRolesPaginator(c.IAM, &iam.ListRolesInput{})
+
+	for paginator.HasMorePages() {
+		page, err := paginator.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("listing IAM roles: %w", err)
+		}
+		for _, role := range page.Roles {
+			if !strings.HasPrefix(*role.RoleName, constants.TestCredentialsStackNamePrefix) {
+				continue
+			}
+			paginator := iam.NewListRoleTagsPaginator(c.IAM, &iam.ListRoleTagsInput{
+				RoleName: role.RoleName,
+			})
+			var tags []types.Tag
+			for paginator.HasMorePages() {
+				page, err := paginator.NextPage(ctx)
+				if err != nil {
+					return nil, fmt.Errorf("listing IAM role tags: %w", err)
+				}
+				tags = append(tags, page.Tags...)
+			}
+			role.Tags = tags
+			if shouldDeleteRole(role, filterInput) {
+				roles = append(roles, *role.RoleName)
+			}
+		}
+	}
+	return roles, nil
+}
+
+func shouldDeleteRole(role types.Role, input FilterInput) bool {
+	var tags []Tag
+	for _, tag := range role.Tags {
+		tags = append(tags, Tag{
+			Key:   *tag.Key,
+			Value: *tag.Value,
+		})
+	}
+	resource := ResourceWithTags{
+		ID:           *role.RoleName,
+		CreationTime: aws.ToTime(role.CreateDate),
+		Tags:         tags,
+	}
+	return shouldDeleteResource(resource, input)
+}
+
+func (c *IAMCleaner) DeleteRole(ctx context.Context, roleName string) error {
+	instanceProfiles, err := c.IAM.ListInstanceProfilesForRole(ctx, &iam.ListInstanceProfilesForRoleInput{
+		RoleName: aws.String(roleName),
+	})
+	if err != nil && errors.IsType(err, &types.NoSuchEntityException{}) {
+		c.Logger.Info("IAM instance profiles already deleted", "roleName", roleName)
+		return nil
```

Comment:
don't we just want to continue and skip the instance profile part but still delete the role?